Add basic install and mount flexvolumes e2e tests

wongma7 opened the pull request:
fixes https://github.com/kubernetes/kubernetes/issues/47010

These two tests install a skeleton "dummy" flex driver, attachable and non-attachable respectively, then test that a pod can successfully use the flex driver. They are labeled disruptive because kubelet and controller-manager get restarted as part of the flex install. IMO it's important to keep this install procedure as part of the test to isolate any bugs with the startup plugin probe code.

There is a bit of an ugly dependency on cluster/gce/config-test.sh because --flex-volume-plugin-dir must be set to a dir that's readable from controller-manager container and writable by the flex e2e test. The default path is not writable on GCE masters with read-only root so I picked a location that looks okay.

In the "dummy" drivers I trick kubelet into thinking there is a mount point by doing "mount -t tmpfs none ${MNTPATH} >/dev/null 2>&1", hope that is okay.

I have only tested on GCE and theoretically they may work on AWS but I don't think there is a need to test on multiple cloudproviders.

-->
```release-note
NONE
```

Commit:
Add basic install and mount flexvolumes e2e tests

wongma7 commented:
/sig storage
@kubernetes/sig-storage-pr-reviews

wongma7 commented:
/test pull-kubernetes-unit
https://github.com/kubernetes/kubernetes/issues/48368

msau42 commented:
cc @verult @chakri-nelluri

msau42 commented on test/e2e/framework/util.go:
```diff
@@ -3641,6 +3641,40 @@ func RestartKubeProxy(host string) error {
+	}
+	var cmd string
+	if ProviderIs("gce") {
+		cmd = "sudo /etc/init.d/kubelet restart"
```
There may also be differences depending on if the image is GCI vs debian

wongma7 replied:
iirc this worked for both but I will double check

wongma7 replied:
oops, GCI is systemd so this doesnt work; how do we check during test if the image is GCI or debian?

wongma7 replied:
nvm, I found NodeOSDistro and MasterOSDistro. Will make use of them.

msau42 commented on test/e2e/framework/util.go:
```diff
@@ -3641,6 +3641,40 @@ func RestartKubeProxy(host string) error {
+}
+
+func WaitForKubeletUp(host string) error {
+	cmd := "curl http://localhost:10248/healthz"
```
Instead of ssh-ing into the host, can you just curl the host directly?

msau42 replied:
Is this port 10248 well defined somewhere?

wongma7 replied:
no, because port is closed probably

10248 is defaulted here: https://github.com/kubernetes/kubernetes/blob/886e04f1fffbb04faf8a9f9ee141143b2684ae68/pkg/apis/componentconfig/v1alpha1/defaults.go#L261 (imo it should be added here, maybe I'll do it in another pr https://github.com/kubernetes/kubernetes/blob/c4eeb238ce19c514b7d3ba719f3311e36101eabe/pkg/master/ports/ports.go)

wongma7 replied:
K, did some more research and 10248 is probably a bad idea to use. Switched to using 10255 for now.

msau42 commented on test/e2e/storage/flexvolume.go:
```diff
@@ -0,0 +1,156 @@
+			},
+			File: "index.html",
+			// Must match content of examples/volumes/flexvolume/dummy(-attachable) domount
+			ExpectedContent: "Hello from flexvolume!",
```
Can you make the content include some generated id, like the test namespace name, so that we can distinguish between different test instances?

wongma7 replied:
it could be even better to make the flex driver name include something generated (it is slightly easier too). e.g. test instance abcd installs driver dummy-abcd, test qwer dummy-qwer, so test instance will be evident from that. And parallel or subsequent tests won't conflict

msau42 commented on test/e2e/storage/flexvolume.go:
```diff
@@ -0,0 +1,156 @@
+	defaultVolumePluginDir = "/usr/libexec/kubernetes/kubelet-plugins/volume/exec"
+	// On gce, root may be read-only and controller-manager containerized. Assume
+	// controller-manager has started with --flex-volume-plugin-dir equal to this
+	// (see cluster/gce/config-test.sh and function mount-master-pd)
```
@verult may be making changes in this area for defining the default flex volume install path for gci images.

ahmetb commented on examples/volumes/flexvolume/dummy-attachable:
```diff
@@ -0,0 +1,104 @@
+#!/bin/sh
+
+# Copyright 2017 The Kubernetes Authors.
```
examples are moved to https://github.com/kubernetes/examples and `examples/` directory soon will be removed. consider submitting this to the examples repository instead.

wongma7 replied:
okay. (I will move them to test/e2e/manifests, not examples/, since atm I'm relying on this script to compile the contents of examples/ https://github.com/kubernetes/kubernetes/blob/master/hack/generate-bindata.sh#L48)

wongma7 commented:
cleaned up and squashed into a single big commit. 

I agonized a bit over how to deal with the half dozen-ish possibly gce distros this could run on but settled on enabling this only for: gci master and debian/gci nodes. I found the alpha suite is only running for gci master and debian nodes anyway.

verult commented on test/e2e/framework/volume_util.go:
```diff
@@ -85,8 +85,10 @@ type VolumeTestConfig struct {
 	WaitForCompletion bool
-	// NodeName to run pod on.  Default is any node.
-	NodeName string
+	// ServerNodeName to run server pod on.  Default is any node.
```
Maybe change the comment to say "NodeName". After reading this, I couldn't make the connection with the "nodeName" from pod spec right away.

verult commented on test/e2e/testing-manifests/flexvolume/dummy-attachable:
```diff
@@ -0,0 +1,104 @@
+
+if [ "$op" = "init" ]; then
+	debug "init $@"
+	log "{\"status\":\"Success\"}"
```
I'd recommend having `capabilities: { attach: true }` here. The default value is mainly for backward compatibility. If it's not in the doc I'll make a PR for it.

verult commented on test/e2e/testing-manifests/flexvolume/dummy-attachable:
```diff
@@ -0,0 +1,104 @@
+	exit 0
+}
+
+domountdevice() {
```
If for whatever reason attach never occurs, can we catch it in this test?

wongma7 replied:
the reader would have to look through controller-manager logs to figure that out.

I will make the test parse the log produced by the flex driver, or at least print it out, so it's more obvious why the test failed

verult replied:
I'm more worried that the test doesn't fail. If k8s only performs mount and skips attach entirely, this driver would still succeed, right?

One thing we could do is keep track of the "attached" state in the driver. Before mounting, if volume is not attached, then don't proceed. Hopefully that'll increase visibility of the error.

wongma7 replied:
Ah, I see. Keeping track of attach state sounds okay. Make the driver closer to reality and easier to reason about.

Alternatively, could simply check in mountdevice that the device name argument is "foo" as returned by attach. Anything else must be a bug.

Fwiw, at the moment the test will fail if attach never occurs. Volume manager would have to be buggy for test to pass like that.

wongma7 replied:
I've added sanity checks to ensure driver fails in these situations (unexpected mount device)

wongma7 commented on test/e2e/framework/util.go:
```diff
@@ -3661,6 +3676,43 @@ func RestartKubeProxy(host string) error {
+	if ProviderIs("gce") && NodeOSDistroIs("debian") {
+		cmd = "sudo /etc/init.d/kubelet restart"
+	} else {
+		cmd = "sudo systemctl restart kubelet"
```
yes, they are on the default kops image

wongma7 commented on cluster/gce/config-test.sh:
```diff
@@ -146,22 +146,34 @@ CONTROLLER_MANAGER_TEST_LOG_LEVEL="${CONTROLLER_MANAGER_TEST_LOG_LEVEL:-$TEST_CL
+# Set flex dir to one that's readable from controller-manager container and writable by the flex e2e test.
+CONTROLLER_MANAGER_TEST_VOLUME_PLUGIN_DIR=""
+if [[ "${MASTER_OS_DISTRIBUTION}" == "gci" ]]; then
+    CONTROLLER_MANAGER_TEST_VOLUME_PLUGIN_DIR="--flex-volume-plugin-dir=/etc/srv/kubernetes/kubelet-plugins/volume/exec"
```
how do you mean, the controller manager container cannot read that dir by default?

/etc/srv/kubernetes/ seems to me to be the least bad place to put it for now since it is used to store various persistent things like auth certs and kubeconfigs.

verult replied:
Sorry yes you are right. I used that directory for kubelet only.

verult commented:
/lgtm

wongma7 commented:
/retest

wongma7 commented:
/assign @gmarek

wongma7 commented:
/assign @verult

wongma7 commented:
pls /approve and re/lgtm. thanks!

verult commented:
/lgtm
/approve

msau42 commented on test/e2e/storage/flexvolume.go:
```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+var _ = framework.KubeDescribe("Flexvolumes [Volume][Disruptive]", func() {
```
You can use SIGDescribe, instead of KubeDescribe.  It will automatically add the [sig-storage] tag to the test name.

wongma7 replied:
fixed. also removed [volume] tag because I don't see it elsewhere anymore

wongma7 commented:
/retest

wongma7 commented:
/retest
https://github.com/kubernetes/kubernetes/issues/50047

wongma7 commented:
@verult please re/lgtm and @gmarek please /approve, you are one of just 3 people who can :) thanks again!

after this merges, I will look into adding it to the alpha test suite.

verult commented:
/lgtm

msau42 commented on test/e2e/storage/flexvolume.go:
```diff
@@ -0,0 +1,207 @@
+	}
+}
+
+var _ = SIGDescribe("Flexvolumes [Disruptive]", func() {
```
It's actually going to run in all the standard, non-alpha suites because there's no [Feature:] tag

gmarek replied:
+1 Please add `[Feature:FlexVolume]` or similar tag.

wongma7 replied:
done 👍

wongma7 commented:
Ah I see, I thought disruptive was enough, I will add a feature tag so we
have more control over where this runs

On Thu, Aug 3, 2017 at 7:18 PM Michelle Au <notifications@github.com> wrote:

> *@msau42* commented on this pull request.
> ------------------------------
>
> In test/e2e/storage/flexvolume.go
> <https://github.com/kubernetes/kubernetes/pull/48366#discussion_r131282189>
> :
>
> > +		}
> +	}
> +	flexDir := path.Join(volumePluginDir, fmt.Sprintf("/%s~%s/", vendor, driver))
> +	return flexDir
> +}
> +
> +func sshAndLog(cmd, host string) {
> +	result, err := framework.SSH(cmd, host, framework.TestContext.
>
> Provider)
> +	framework.LogSSHResult(result)
> +	framework.ExpectNoError(err)
> +	if result.Code != 0 {
> +		framework.Failf("%s returned non-zero, stderr: %s", cmd, result.Stderr)
> +	}
> +}
> +
> +var _ = SIGDescribe("Flexvolumes [Disruptive]", func() {
>
> It's actually going to run in all the standard, non-alpha suites because
> there's no [Feature:] tag
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/kubernetes/pull/48366#pullrequestreview-54247972>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMgP-ETsCvVIFBIST9qGtLT-p9vxTsoGks5sUlVbgaJpZM4OLFv9>
> .
>

wongma7 commented:
Or we can merge as is, flex isn’t actually gated by alpha is it. Anyway
I’ll look tomorrow

On Thu, Aug 3, 2017 at 7:36 PM Matthew Wong <mawong@redhat.com> wrote:

> Ah I see, I thought disruptive was enough, I will add a feature tag so we
> have more control over where this runs
>
> On Thu, Aug 3, 2017 at 7:18 PM Michelle Au <notifications@github.com>
> wrote:
>
>> *@msau42* commented on this pull request.
>> ------------------------------
>>
>> In test/e2e/storage/flexvolume.go
>> <https://github.com/kubernetes/kubernetes/pull/48366#discussion_r131282189>
>> :
>>
>> > +		}
>> +	}
>> +	flexDir := path.Join(volumePluginDir, fmt.Sprintf("/%s~%s/", vendor, driver))
>> +	return flexDir
>> +}
>> +
>> +func sshAndLog(cmd, host string) {
>> +	result, err := framework.SSH(cmd, host, framework.TestContext.
>>
>> Provider)
>> +	framework.LogSSHResult(result)
>> +	framework.ExpectNoError(err)
>> +	if result.Code != 0 {
>> +		framework.Failf("%s returned non-zero, stderr: %s", cmd, result.Stderr)
>> +	}
>> +}
>> +
>> +var _ = SIGDescribe("Flexvolumes [Disruptive]", func() {
>>
>> It's actually going to run in all the standard, non-alpha suites because
>> there's no [Feature:] tag
>>
>> —
>> You are receiving this because you were mentioned.
>> Reply to this email directly, view it on GitHub
>> <https://github.com/kubernetes/kubernetes/pull/48366#pullrequestreview-54247972>,
>> or mute the thread
>> <https://github.com/notifications/unsubscribe-auth/AMgP-ETsCvVIFBIST9qGtLT-p9vxTsoGks5sUlVbgaJpZM4OLFv9>
>> .
>>
>

gmarek commented:
/lgtm

gmarek commented:
/retest

wongma7 commented:
/retest
    Error from server: error dialing backend: dial tcp 10.128.0.5:10250: getsockopt: connection refused

wongma7 commented:
/retest

msau42 commented:
@wongma7 I'm seeing a similar issue with the local storage tests.  I think something changed in the test env and it cannot resolve the node hostnames anymore.

msau42 commented:
Oh wait nevermind, you're referring to etcd3 test, not the flex tests.  Well, actually, you may run into a similar issue with the flex tests, since you're trying to SSH Into the nodes.
